Decouple session from connection

lutovich opened the pull request.

lutovich committed:
Simplify driver implementations

Previously there existed two driver implementations `DirectDriver` and
`RoutingDriver`. Main difference was that first one used connection pool
directly while the second one used load balancing connection pool.

This commit introduces an interface to unify connection pool and load balancer
and collapses two driver implementations into one that uses this interface.
This also makes it more convenient to decouple connection from session because
session can now use this same single interface without knowing if it is a
direct or routing session.

lutovich committed:
Makes NetworkSession aware of ConnectionProvider

So it can use the provider to acquire connections when needed. This is a piece
of preparatory work to decouple session from connection. This commit does not
change how sessions use connections.

lutovich committed:
Remove unused field from InternalStatementResult

lutovich committed:
Decouple session from connection

Previously session was bound to a connection and used this same connection
throughout its lifetime. This means that there was no load-balancing within
single session and that connection could have been claimed by the session
for a long time without actually being used.

This commit makes session obtain new connections when needed and return them
to the pool as soon as possible. New connection is obtained for:
 * each `Session#run()`
 * each `Session#beginTransaction()` and created transaction holds same
 connection throughout its lifetime
Connection is returned back to the pool when either:
 * result is fully read into memory
 * transaction is closed
 * subsequent `Session#run()` is invoked

Introduced new interface `ConnectionHandler` to manage connections around
session, result and transaction. `NetworkSession` is the only implementation
of this interface.
Also added new `DriverClosedException` thrown on every interaction with a
closed driver.

lutovich committed:
Simplify logging around sessions and connections

Hash codes instead of UUID strings for session/connection identification,
removed `Connection#logger()` interface method.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/InternalDriver.java:
```diff
@@ -61,42 +64,57 @@ public final Session session()
             // For 2. this closeResources will close the new connection pool just created to ensure no resource leak.
             closeResources();
-            throw driverCloseException();
+            throw new DriverClosedException();
```
Maybe not a new exception type in driver for now

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/InternalStatementResult.java:
```diff
@@ -319,6 +321,7 @@ private boolean tryFetchNext()
         {
             if ( done )
             {
+                connectionHandler.resultBuffered();
```
if you do `connectionHandler.resultBuffered()` at the place where `done` is set to be true, then you probably could also address failure/ignore cases too.

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java:
```diff
@@ -132,21 +118,23 @@ public synchronized void reset()
-        connection.resetAsync();
+        if ( currentConnection != null )
+        {
+            currentConnection.resetAsync();
```
I would suggest to remove reset. Make this method deprecated and then give an empty impl inside.

lutovich replied:
Can we maybe remove it in a separate PR? This one is quite big already :)

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/NetworkSession.java:
```diff
@@ -241,25 +203,42 @@ public TypeSystem typeSystem()
+        if ( currentTransaction != null && currentTransaction == tx )
+        {
+            lastBookmark = currentTransaction.bookmark();
+            currentTransaction = null;
```
connection also need to be closed

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/net/pooling/PooledSocketConnection.java:
```diff
@@ -58,7 +59,7 @@
 
     private boolean unrecoverableErrorsOccurred = false;
-    private Runnable onError = null;
+    private ConnectionHandler connectionHandler;
```
suggesting rename something else such as `ConnectionTerminationHandler`?

lutovich committed:
More unit tests for NetworkSession

To check that it handles connections properly.

lutovich committed:
Renamed ConnectionHandler to SessionResourcesHandler

And improved method names to better reflect the purpose of the interface.

lutovich committed:
Removed DriverClosedException

All usages changed back to IllegalStateException. This revert is done because
we should first refine driver errors hierarchy and only then introduce
specific exception types.

lutovich committed:
Close connection after result reading failure

When result fetching fails (cypher runtime error or network error) exception
is thrown but it is still possible to maintain a reference to the result
object. It'll just be empty after failure. It is not possible to fetch any
other records after failure and the underlying pooled connection can be
closed and returned to the pool.

This commit makes result return its connection to the pool on failure. This
connection (if healthy) can be used by subsequent sessions. Also added tests
for connection handling.

zhenlineo commented on driver/src/test/java/org/neo4j/driver/internal/InternalStatementResultTest.java:
```diff
@@ -386,23 +386,89 @@ public void shouldNotPeekIntoTheFutureWhenResultIsEmpty()
+        SessionResourcesHandler resourcesHandler = mock( SessionResourcesHandler.class );
+        StatementResult result = createResult( 10, resourcesHandler );
+
+        assertNotNull( result.summary() );
```
maybe summary several times to ensure only called once?
Similar to other methods and also some combination of all the methods? such as result.single(), result.summary(), then result.consume()?

lutovich replied:
Added a separate test for this here: https://github.com/neo4j/neo4j-java-driver/pull/324/commits/0ddac3d5036b7017f80f6592ba4815b0b00bcd87

lutovich committed:
Test for fetching result multiple times

lutovich committed:
Sync only open connections

Added a check to sync only open connections before closing them in the session.
